shell/browser/net/resolve_host_helper.cc

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2018 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/net/resolve_host_helper.h"
+
+#include <utility>
+#include <vector>
+
+#include "base/functional/bind.h"
+#include "base/values.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/storage_partition.h"
+#include "net/base/host_port_pair.h"
+#include "net/base/net_errors.h"
+#include "net/base/network_isolation_key.h"
+#include "net/dns/public/resolve_error_info.h"
+#include "shell/browser/electron_browser_context.h"
+#include "url/origin.h"
+
+using content::BrowserThread;
+
+namespace electron {
+
+ResolveHostHelper::ResolveHostHelper(ElectronBrowserContext* browser_context)
+    : browser_context_(browser_context) {}
+
+ResolveHostHelper::~ResolveHostHelper() {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  DCHECK(!owned_self_);
+  DCHECK(!receiver_.is_bound());
+  // Clear all pending requests if the HostService is still alive.
+  pending_requests_.clear();
+}
+
+void ResolveHostHelper::ResolveHost(std::string host,
+                                    ResolveHostCallback callback) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  // Enqueue the pending request.
+  pending_requests_.emplace_back(host, std::move(callback));
```

Comment:
dns_api.cc doesn't implement request serialization like this. is there a particular reason we need this extra complexity?

Reply:
I think we need to either allocate one helper per each request or do queuing like here. Now that you say it I wonder if there's potentially a race in the code. I'll change the pattern to always allocate it.

Reply:
Done.